lib/src/protocol/kawa_h1/mod.rs

```diff
@@ -0,0 +1,1933 @@
+    }
+
+    pub fn set_backend_id(&mut self, backend_id: String) {
+        self.backend_id = Some(backend_id);
+    }
+
+    pub fn set_backend_token(&mut self, token: Token) {
+        self.backend_token = Some(token);
+    }
+
+    pub fn clear_backend_token(&mut self) {
+        self.backend_token = None;
+    }
+
+    pub fn cancel_backend_timeout(&mut self) {
+        self.container_backend_timeout.cancel();
+    }
+
+    pub fn set_backend_timeout(&mut self, dur: Duration) {
+        if let Some(token) = self.backend_token.as_ref() {
+            self.container_backend_timeout.set_duration(dur);
+            self.container_backend_timeout.set(*token);
+        }
+    }
+
+    pub fn front_socket(&self) -> &TcpStream {
+        self.frontend_socket.socket_ref()
+    }
+
+    /// WARNING: this function removes the backend entry in the session manager
+    /// IF the backend_token is set, so that entry can be reused for new backend.
+    /// I don't think this is a good idea, but it is a quick fix
+    fn close_backend(&mut self, proxy: Rc<RefCell<dyn L7Proxy>>, metrics: &mut SessionMetrics) {
+        debug!(
+            "{}\tPROXY [{}->{}] CLOSED BACKEND",
+            self.log_context(),
+            self.frontend_token.0,
+            self.backend_token
+                .map(|t| format!("{}", t.0))
+                .unwrap_or_else(|| "-".to_string())
+        );
+
+        let proxy = proxy.borrow();
+        if let Some(socket) = &mut self.backend_socket.take() {
+            if let Err(e) = proxy.deregister_socket(socket) {
+                error!("error deregistering back socket({:?}): {:?}", socket, e);
+            }
+            if let Err(e) = socket.shutdown(Shutdown::Both) {
+                if e.kind() != ErrorKind::NotConnected {
+                    error!("error shutting down back socket({:?}): {:?}", socket, e);
+                }
+            }
+        }
+
+        if let Some(token) = self.backend_token {
+            proxy.remove_session(token);
+
+            if self.backend_connection_status != BackendConnectionStatus::NotConnected {
+                self.backend_readiness.event = Ready::EMPTY;
+            }
+
+            if self.backend_connection_status == BackendConnectionStatus::Connected {
+                gauge_add!("backend.connections", -1);
+                gauge_add!(
+                    "connections_per_backend",
+                    -1,
+                    self.cluster_id.as_deref(),
+                    metrics.backend_id.as_deref()
+                );
+            }
+
+            self.set_backend_connected(BackendConnectionStatus::NotConnected, metrics);
+
+            if let Some(backend) = self.backend.take() {
+                self.clear_backend_token();
+                backend.borrow_mut().dec_connections();
+                self.cancel_backend_timeout();
+            }
+        }
+    }
+
+    fn check_circuit_breaker(&mut self) -> anyhow::Result<()> {
```

Comment:
please write a documenting comment like "Checks the number of connection attempts against authorized connection retries"